### PR TITLE
Add floating-point optimizer based on Icing

### PR DESCRIPTION
**Pull request by HeikoBecker**
This PR contains a CakeML source floating-point optimizer that extends CakeML with
- a real-valued semantics for floating-point verification
- a nondeterministic (oracle-based) floating-point semantics that applies fast-math-style optimizations
- a optimizer in the directory `icing` that rigorously optimizes floating-point kernels using fast-math-style optimizations

To infer floating-point error bounds, the PR also adds the FloVer certificate checker in the `icing` directory.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -75,4 +75,9 @@ compiler/benchmarks/sml/polyc_*
 compiler/benchmarks/sml/sml_*
 compiler/benchmarks/*.dat
 compiler/benchmarks/*.eps
+<<<<<<< HEAD
```
Comment (HeikoBecker) [inline .gitignore:78]: Remove this.

File: semantics/Holmakefile [hunk 2]
```diff
@@ -7,6 +7,7 @@ all: $(DEFAULT_TARGETS) README.md
 .PHONY: all
 
 README_SOURCES = grammar.txt astPP.sml $(wildcard *Script.sml) $(wildcard *Lib.sml) $(wildcard *Syntax.sml)
+
```
Comment (HeikoBecker) [inline semantics/Holmakefile:10]: Remove this line.

File: semantics/astScript.sml [hunk 3]
```diff
@@ -68,9 +67,19 @@ Datatype:
   | Equality
   (* FP operations *)
   | FP_cmp fp_cmp
+  (* | FP_pred of fp_pred *)
```
Comment (HeikoBecker) [inline semantics/astScript.sml:70]: Remove this.

File: semantics/astScript.sml [hunk 4]
```diff
@@ -127,6 +136,44 @@ Datatype:
...
+  | _ => Simple
+)))`;
+
+
+(*val isFpBool: op -> bool*)
+val _ = Define `
+ ((isFpBool:op -> bool) op=
+   ((case op of
+    FP_cmp _ => T
+  (* | FP_pred _ -> true *)
```
Comment (HeikoBecker) [inline semantics/astScript.sml:172]: Remove this.

File: semantics/astScript.sml [hunk 5]
```diff
@@ -127,6 +136,44 @@ Datatype:
...
+ op_class =
+    EvalOp (* Eval primitive *)
+  | FunApp (* function application *)
+  | Simple (* arithmetic operation, no finite-precision/reals *)
+  | Icing (* 64-bit floating-points *)
+  | Reals (* real numbers *)
+End
+
+(*val getOpClass : op -> op_class*)
+val _ = Define `
```
Comment (HeikoBecker) [inline semantics/astScript.sml:150]: Use Definition syntax here, remove type comment.s

File: semantics/fpValTreeScript.sml [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+       Fp_const word64
+     | Fp_uop fp_uop fp_word_val
+     | Fp_bop fp_bop fp_word_val fp_word_val
+     | Fp_top fp_top fp_word_val fp_word_val fp_word_val
+     | Fp_wopt fp_opt fp_word_val
+End
+
+Datatype:
+ fp_bool_val =
+     (* | Fp_pred of fp_pred * fp_word_val *)
```
Comment (HeikoBecker) [inline semantics/fpValTreeScript.sml:48]: Remove this.

File: semantics/fpOptScript.sml [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+*)
+
+Datatype:
+ fp_pat =
+       Word  word64
+     | PatVar  num
+     | Unop  fp_uop  fp_pat
+     | Binop  fp_bop  fp_pat  fp_pat
+     | Terop  fp_top  fp_pat  fp_pat  fp_pat
+     (* | Pred  fp_pred * fp_pat *)
```
Comment (HeikoBecker) [inline semantics/fpOptScript.sml:22]: Remove.

File: semantics/semanticPrimitivesScript.sml [hunk 8]
```diff
@@ -155,6 +172,30 @@ Type compiler_args = ``: ((num # num) # v # dec list)``
...
+ | FPScope fp_opt (* currently under scope *)
+End
+
+Datatype:
+ fpState =
+  <| rws: fp_rw list
+   ; opts: num -> rewrite_app list
+   ; choices: num
+   ; canOpt : optChoice
+   ; assertions : num -> v store -> v sem_env -> bool
```
Comment (HeikoBecker) [inline semantics/semanticPrimitivesScript.sml:194]: Remove this.

File: semantics/semanticPrimitivesScript.sml [hunk 9]
```diff
@@ -647,6 +736,48 @@ End
      (Conv (SOME (TypeStamp "Fp_less" fp_cmp_type_num)) []))`;
 
 
+(*val enc_real_uop : real_uop -> v*)
```
Comment (HeikoBecker) [inline semantics/semanticPrimitivesScript.sml:739]: Remove the encoding part here.

**Comment by HeikoBecker** [comment]
All of the big conceptual issues should be fixed now. This should be good to go as soon as the regression tests pass.

**Review by oskarabrahamsson — CHANGES_REQUESTED** [review]
Not a fan of custom special-use tactics, esp. in this case when they’re so short.

File: compiler/repl/evaluate_initScript.sml [hunk 10]
```diff
@@ -635,23 +664,60 @@ Proof
...
+    gvs[do_app_cases, v_ok_thm]
+    \\ ‘s with <| refs := s.refs; ffi := s.ffi |> = s’ suffices_by gs[]
+    \\ gs[state_component_equality])
   \\ Cases_on ‘op’ \\ gs []
 QED
 
 Theorem do_app_ok = SIMP_RULE (srw_ss()) [LET_THM] do_app_ok;
 
+
+val icing_rename_tac =
```
Comment (oskarabrahamsson) [inline compiler/repl/evaluate_initScript.sml:688]: Not a fan of custom special-use tactics, esp. in this case when they’re so short.

**Review by oskarabrahamsson — APPROVED** [review]

**Pull request merged**